encoding/protobuf/encoding_integration_test.go

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package protobuf_test
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/encoding/protobuf/internal/testpb"
+	"go.uber.org/yarpc/transport/grpc"
+	"go.uber.org/yarpc/yarpctest"
+)
+
+func TestProtobufEncoding(t *testing.T) {
+	client, cleanup := initalizeClientAndServer(t, &encodingIntegrationTestServer{})
+
+	defer cleanup()
+	for _, test := range []struct {
+		desc            string
+		request         *yarpctest.Call
+		expectedHeaders map[string]string
+	}{
+		{
+			desc: "test1",
+			request: &yarpctest.Call{
+				Procedure: "ABC1",
+			},
+			expectedHeaders: map[string]string{
+				"CallerProcedure": "ABC1",
+				"Procedure":       "uber.yarpc.encoding.protobuf.Test::Unary",
+			},
+		},
+		{
+			desc:    "test2",
+			request: &yarpctest.Call{},
+			expectedHeaders: map[string]string{
+				"CallerProcedure": "",
+				"Procedure":       "uber.yarpc.encoding.protobuf.Test::Unary",
+			},
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			ctx = yarpctest.ContextWithCall(ctx, test.request)
+			defer cancel()
+
+			headersStr := mapToStr(test.expectedHeaders)
+			_, err := client.Unary(ctx, &testpb.TestMessage{Value: headersStr})
```

Comment:
Should this call return a response where the message contains the value of the headers of the request? Basically you would do this in the handler:

```
func (encodingIntegrationTestServer) Unary(ctx context.Context, msg *testpb.TestMessage) (*testpb.TestMessage, error) {
rep := &testpb.TestMessage{}
rep.Value := ctx.Headers // stringily this
return rep
}
```

Then in the test itself you could simply do:
```
assert.Equal(t, tt.expectedHeader, resp.Value) // you would need to have the same format stringvsmap of course
```

What do you think? This would simplify a lot the server/client code. And all check on the test would be self contained in the test itself and not spread to different functions

Note: we should rename `encodingIntegrationTestServer` to something like `headerIntegrationTestServer`

Reply:
Moving these integration tests to the separate pull request and I will take care of this in that